File: frontend/src/app/services/savings-goal.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { CreateSavingsGoalDto, SavingsGoal } from '../models/savings-goal.model';

@Injectable({
  providedIn: 'root'
})
export class SavingsGoalService {

  private apiUrl = 'http://localhost:8081/api/savings-goals';

  constructor(private http: HttpClient) {}

  // Récupérer tous les savings goals de l'utilisateur connecté
  getAllSavingsGoals(): Observable<SavingsGoal[]> {
    return this.http.get<SavingsGoal[]>(this.apiUrl);
  }

  // Récupérer un savings goal par ID
  getSavingsGoalById(id: number): Observable<SavingsGoal> {
    return this.http.get<SavingsGoal>(`${this.apiUrl}/${id}`);
  }

  // Créer un nouveau savings goal
  createSavingsGoal(goal: CreateSavingsGoalDto): Observable<SavingsGoal> {
    return this.http.post<SavingsGoal>(this.apiUrl, goal);
  }

  // Modifier un savings goal
  updateSavingsGoal(id: number, goal: Partial<CreateSavingsGoalDto>): Observable<SavingsGoal> {
    return this.http.put<SavingsGoal>(`${this.apiUrl}/${id}`, goal);
  }

  // Supprimer un savings goal
  deleteSavingsGoal(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}

File: backend/src/main/java/com/walletly/walletly_backend/service/SavingsGoalService.java
package com.walletly.walletly_backend.service;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.walletly.walletly_backend.exception.ErrorMessages;
import com.walletly.walletly_backend.exception.NotFoundException;
import com.walletly.walletly_backend.model.SavingsGoal;
import com.walletly.walletly_backend.model.User;
import com.walletly.walletly_backend.repository.SavingsGoalRepository;

@Service
public class SavingsGoalService {

    @Autowired
    private SavingsGoalRepository savingsGoalRepository;

    @Autowired
    private AuthService authService;

    @Autowired
    private InputSanitizer inputSanitizer;

    /**
     * Créer un nouveau savings goal pour un utilisateur
     */
    public SavingsGoal createSavingsGoal(SavingsGoal savingsGoal, Long userId) {
        User user = authService.getUserById(userId);
        savingsGoal.setName(inputSanitizer.sanitizePlainText(savingsGoal.getName(), "savingsGoal.name"));
        savingsGoal.setUser(user);
        return savingsGoalRepository.save(savingsGoal);
    }

    /**
     * Récupérer tous les savings goals d'un utilisateur
     */
    public List<SavingsGoal> getAllByUser(Long userId) {
        return savingsGoalRepository.findByUser_Id(userId);
    }

    /**
     * Récupérer un savings goal par ID
     */
    public SavingsGoal getById(Long id, Long userId) {
        return savingsGoalRepository.findByIdAndUser_Id(id, userId)
            .orElseThrow(() -> new NotFoundException(ErrorMessages.SAVINGS_GOAL_NOT_FOUND));
    }

    /**
     * Modifier un savings goal (nom, targetAmount, currentAmount, deadline)
     */
    public SavingsGoal updateSavingsGoal(Long id, SavingsGoal updated, Long userId) {
        SavingsGoal existing = getById(id, userId);

        existing.setName(inputSanitizer.sanitizePlainText(updated.getName(), "savingsGoal.name"));
        existing.setTargetAmount(updated.getTargetAmount());
        existing.setCurrentAmount(updated.getCurrentAmount());
        existing.setDeadline(updated.getDeadline());

        return savingsGoalRepository.save(existing);
    }

    /**
     * Supprimer un savings goal
     */
    public void deleteSavingsGoal(Long id, Long userId) {
        SavingsGoal existing = getById(id, userId);

        savingsGoalRepository.delete(existing);
    }
}
